### ISSUE TITLE
Environment Variable strangeness

### ISSUE THREAD

**Issue by mvastola** [buildkite-plugins/ecr-buildkite-plugin#72]
I should start off by saying I'm not sure if this belongs here or in the docker plugin repo, but here is my best guess.

I am running into a very strange issues using the env vars set by this plugin (namely `${DOCKER_CONFIG}`)  in my pipeline.

Judging by the fact no one else appears to have had this issue<strong>**</strong>, I'm guessing that mine is an unusual use case, but what I'm trying to do is mount my docker credentials from ECR into an instance.

By all appearances, a build step like this should work fine:
```yaml
- name: ":docker: Building image"
    plugins:
      - improbable-eng/metahook:
          pre-command: |
            echo $DOCKER_CONFIG
            echo \$DOCKER_CONFIG
      - ecr#v2.4.0:
          login: true
          account-ids: '[ACCOUNT NUMBER]'
          region: us-east-1
          no-include-email: true
          assume_role:
            role_arn: "arn:aws:iam::[ACCOUNT NUMBER]:role/[ROLE NAME]"
      - docker#v3.8.0:
          image: moby/buildkit:master
          privileged: true
          userns: host
          always-pull: true
          propagate-environment: true
          mount-checkout: true
          mount-buildkite-agent: false
          tmpfs:
            - '/tmp'
          volumes:
            - '${HOME}/.aws:/home/user/.aws'
            - '${DOCKER_CONFIG}:/home/user/.docker'
          environment:
            - 'DOCKER_CONFIG=/home/user/.docker'
          tty: true
          debug: true
          entrypoint: buildctl-daemonless.sh
          command: [ 'build', '...more args here' ]

```
(Note: the `improbable-eng/metahook` plugin is for debugging and not needed to reproduce.)

What seems to be happening (and I have precisely zero guesses as to why) is that `DOCKER_CONFIG` has two different values: it holds the path to one temporary folder when the variables in the `pipeline.yml` are interpolated, and is reset somewhere to an entirely _different_ folder when the command is invoked. 

The result is that the `improbable-eng/metahook` `pre-command` in my build yields two completely different values, as the command appears to undergo expansion twice. 

After spending _far_ too much time trying to figure out what was going on here, I noticed this in the docs:
https://buildkite.com/docs/pipelines/secrets#anti-pattern-referencing-secrets-in-your-pipeline-yaml. (Which was not helpfully named, given I was just trying to reference a filesystem path.)

In any case, the suggestion in the documentation did not work: changing the volume mapping to '$${DOCKER_CONFIG}:/home/user/.docker' results in it becoming (the literal) `${DOCKER_CONFIG}` when `docker run` is executed, yielding the error:

> docker: Error response from daemon: create ${DOCKER_CONFIG}: "${DOCKER_CONFIG}" includes invalid characters for a local volume name, only "[a-zA-Z0-9][a-zA-Z0-9_.-]" are allowed. If you intended to pass a host directory, use absolute path.

So I'm just really confused.
- Why is `${DOCKER_CONFIG}` getting changed when it's already been set?
- Why can I literally just reference an environment variable (set in the eponymous `environment` stage) in what is both a later stage AND a plugin defined after the ECR one?
- If this _isn't_ a bug (which would be upsetting), how would I work around it? The barriers I see are:

    1. The nature of the `docker` plugin is such that there's no ability to run commands outside the container in the same step
    2. I can't easily store them in an earlier stage because that isn't guaranteed to be the same EC2 instance (and an artifact would be insecure)
    3. A workaround along the lines of `improbable-eng/metahook` seems a bit hacky.

--------

<strong>**</strong> This may actually be the underlying cause of #32, but I'm not sure and that issue is stale (2+ years old), so I'm creating this new one.

**Comment by keithduncan** [comment]
Sorry this gave you trouble @mvastola and broke expectations of how things should work. I can provide some details on what is happening here.

> * Why is `${DOCKER_CONFIG}` getting changed when it's already been set?

I’m assuming you’re using the Elastic CI Stack for AWS. Here, an [Agent Location Job Environment Hook](https://github.com/buildkite/elastic-ci-stack-for-aws/blob/c0371d65388570e95d74c855178a1a262e47d5ab/packer/linux/conf/buildkite-agent/hooks/environment#L8-L11) sets `DOCKER_CONFIG` to a per-job temporary directory, and cleans it up in a [pre-exit](https://github.com/buildkite/elastic-ci-stack-for-aws/blob/c0371d65388570e95d74c855178a1a262e47d5ab/packer/linux/conf/buildkite-agent/hooks/pre-exit#L5-L8) hook.

This is done to avoid leaking state between jobs, and to better support multiple agents per instance which would otherwise all be sharing and modifying the same config.

That is why `DOCKER_CONFIG` will have one value at pipeline upload interpolation time, and another when the job is actually run.

> Why can I literally just reference an environment variable (set in the eponymous environment stage) in what is both a later stage AND a plugin defined after the ECR one?

I’m afraid I’m not following you here, can you rephrase this question?

> * If this _isn't_ a bug (which would be upsetting), how would I work around it? The barriers I see are:

It may be viable to update the docker plug-in per your issue https://github.com/buildkite-plugins/docker-buildkite-plugin/issues/189 to expand variables 🤔 there is already some [rudimentary expansion](https://github.com/buildkite-plugins/docker-buildkite-plugin/blob/a756a5f5e76423c99fc0649c1d7b0fd6ea78aa9c/hooks/command#L126) but it would be delicate to avoid introducing a command injection attack vector.

Finally, I know this may not be the answer you want 😅 but it may be that this use case is deemed outside those we aim to cover with the docker plug-in. If that is the case, some workarounds for this would be:

- Eschew the docker plug-in and author a command script that does exactly what you want in this pipeline’s repository, and invoke it from the step
- If used in more than one pipeline, share this command script across them using a private plug-in
